### PR TITLE
Remove `with*` and make `with` treat `_` as special

### LINKED ISSUE #1993: `with` forms cannot parse multiple non-named with-exprs
Currently, `with` parses like so (simplification mine):
```
=> (with [a b c d e] ...)
with b as a:
  with d as c:
    with e:
      ...
```

As a result, there is no way of writing multiple non-named forms, nor is there any way to have a non-named form anywhere other than the end.

I propose the following instead, with a set of outer brackets for the with-exprs and inner brackets for any named with-exprs:

```
=> (with [[a b] c d [e f]] ...)
with b as a, c, d, f as e:
  ...
```

### PR DESCRIPTION
**Pull request by peaceamongworlds**
Fixes #1993.

`with*` is deprecated and `with` is now a special form. It also compile `_` to unnamed context managers.

### CODE REVIEW

File: hy/compiler.py
Comment (scauligi) [inline hy/compiler.py:907]: Did you mean to leave these commented?

File: tests/native_tests/contrib/walk.hy
Comment (scauligi) [inline tests/native_tests/contrib/walk.hy:62]: These new tests seems unrelated to the current PR

**Review by scauligi — COMMENTED** [review]
Can you add tests for the new `with` pattern(s)? In particular we want to make sure that using `_` in fact makes an unnamed manager instead of introducing a new `_` name.

File: tests/native_tests/contrib/walk.hy
Comment (peaceamongworlds) [reply tests/native_tests/contrib/walk.hy:62]: The old tests used `with`, which is no longer a macro, so I had to change them to something else.

File: hy/compiler.py
Comment (peaceamongworlds) [reply hy/compiler.py:907]: No, that was silly

**Comment by Kodiologist** [comment]
It looks like there was a mistake in the original model pattern for `with*`, where `FORM` was used instead of `SYM` for the new variable. Might as well fix that.

**Comment by peaceamongworlds** [comment]
> It looks like there was a mistake in the original model pattern for `with*`, where `FORM` was used instead of `SYM` for the new variable. Might as well fix that.

Python allows for unpacking in `with` statements if the context manager returns a tuple or list. For example:
```python
[ins] In [134]: class Test():
           ...:     def __init__(self, x, y):
           ...:         self.x = x
           ...:         self.y = y
           ...:     def __enter__(self):
           ...:         return self.x, self.y
           ...:     def __exit__(self, type, value, traceback):
           ...:         pass
[ins] In [136]: with Test(1, 2) as (x, y):
           ...:     print(x, y)
           ...: 
1 2
```
So I think that `FORM` is correct.

**Comment by Kodiologist** [comment]
Why not merge it, then? It has two approvals.

File: hy/compiler.py
Comment (scauligi) [inline hy/compiler.py:900]: minor nit: use `Inf` instead of `float("inf")` to match other patterns

File: docs/api.rst
Comment (peaceamongworlds) [reply docs/api.rst]: Done

File: docs/api.rst
Comment (peaceamongworlds) [reply docs/api.rst]: Fixed

**Comment by peaceamongworlds** [comment]
There's something weird going on with `oneplus`. There was a bug in it's implementation, see https://github.com/vlasovskikh/funcparserlib/commit/10f4017287b88fc85a67a6e13760f915a137a8e4, which has since been fixed. However, pip (and pacman for that matter) seem to install the old version which still has the bug.

**Comment by Kodiologist** [comment]
I posted an issue requesting a new release of funcparserlib, so hopefully that will be taken care of soon. Failing that, you can try making Hy depend on an unreleased version from GitHub.

**Comment by Kodiologist** [comment]
Or we can switch back to `times` for now. That's fair, too.

**Comment by peaceamongworlds** [comment]
I think `times` is fine for now. When funcparserlib is updated, it'll be easy to update this as well.

**Review by scauligi — APPROVED** [review]

**Pull request merged**
